# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn import preprocessing

# Load in Data

The dataset used is available in the GitHub repository under the name **"credit_card_churn.csv**. This data was retrieved from Kaggle.com, with a direct link located in the repository README file.

In [2]:
#Use pandas, load in dataset 
credit_df = pd.read_csv("credit_card_churn.csv")

#Check that data loaded in correctly
credit_df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

# Understanding the Data

Before going into the deep analysis, it is essential to understand the data that will be used in the modeling solution. Looking at the properties this data set will provide enough information to determine if a solution to the customer attrition issue can be produced.

In [3]:
#credit_df.info()

There are no nulls present in any of the variables, each has 10127 observations - indicating no need to alter the dataset due to missing information. There are 23 columns, 10127 observations, and no null values. Some of these columns will not be necessary for the later stages of analysis and can be subsequently cut from our working dataframe. 



To get a better idea of what can be cut, here are the variable names and definitions:
1. **CLIENTNUM**: *Client number. Unique identified for the customer holding the account*
2. **Attrition_Flag**: *Internal event (customer activity) variable - if the account is closed then 1 else 0* ***TARGET VARIABLE***
3. **Customer_Age**: *Demographic variable - Customer's Age in Years*
4. **Gender**: *Demographic variable - M=Male, F=Female*
5. **Dependent_count**: *Demographic variable - Number of dependents*
6. **Education_Level**: *Demographic variable - Educational Qualification of the account holder*
7. **Marital_Status**: *Demographic variable - Married, Single, Divorced, Unknown*
8. **Income_Category**: *Demographic variable - Annual Income Category of the account holder*
9. **Card_Category**: *Product variable - Type of card*
10. **Months_on_book**: *Period of relationship with bank*
11. **Total_Relationship_Count**: *Total number of products held by the customer*
12. **Months_Inactive_12_mon**: *Number of months inactive in the last 12 months*
13. **Contacts_Count_12_mon**: *Number of contacts in the last 12 months*
14. **Credit_Limit**: *Credit limit on the Credit Card*
15. **Total_Revolving_Bal**: *Total revolving balance on the Credit Card*
16. **Avg_Open_To_Buy**: *Open to Buy Credit Line (Average of last 12 months)*
17. **Total_Amt_Chng_Q4_Q1**: *Change in Transaction Amount (Q4 over Q1)*
18. **Total_Trans_Amt**: *Total transaction amount (last 12 months)*
19. **Total_Trans_Ct**: *Total transaction count (last 12 months)*
20. **Total_Ct_Chng_Q4_Q1**: *Change in transaction count (Q4 over Q1)*
21. **Avg_Utilization_Ratio**: *Average card utilization ratio*
22. **Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2**: *Naive Bayes - note from author to remove*
23. **Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1** *Naive Bayes - note from author to remove*






Before beginning the cleaning process, it is beneficial to be familiar with the descriptive statistics of the numerical variables. 

In [4]:
#Use the describe() function to evaluate descriptive stats
credit_df.describe()

CLIENTNUM  Customer_Age  Dependent_count  Months_on_book  \
count  1.012700e+04  10127.000000     10127.000000    10127.000000   
mean   7.391776e+08     46.325960         2.346203       35.928409   
std    3.690378e+07      8.016814         1.298908        7.986416   
min    7.080821e+08     26.000000         0.000000       13.000000   
25%    7.130368e+08     41.000000         1.000000       31.000000   
50%    7.179264e+08     46.000000         2.000000       36.000000   
75%    7.731435e+08     52.000000         3.000000       40.000000   
max    8.283431e+08     73.000000         5.000000       56.000000   

       Total_Relationship_Count  Months_Inactive_12_mon  \
count              10127.000000            10127.000000   
mean                   3.812580                2.341167   
std                    1.554408                1.010622   
min                    1.000000                0.000000   
25%                    3.000000                2.000000   
50%                    4.000000                2.000000   
75%                    5.000000                3.000000   
max                    6.000000                6.000000   

       Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
count           10127.000000  10127.000000         10127.000000   
mean                2.455317   8631.953698          1162.814061   
std                 1.106225   9088.776650           814.987335   
min                 0.000000   1438.300000             0.000000   
25%                 2.000000   2555.000000           359.000000   
50%                 2.000000   4549.000000          1276.000000   
75%                 3.000000  11067.500000          1784.000000   
max                 6.000000  34516.000000          2517.000000   

       Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
count     10127.000000          10127.000000     10127.000000    10127.000000   
mean       7469.139637              0.759941      4404.086304       64.858695   
std        9090.685324              0.219207      3397.129254       23.472570   
min           3.000000              0.000000       510.000000       10.000000   
25%        1324.500000              0.631000      2155.500000       45.000000   
50%        3474.000000              0.736000      3899.000000       67.000000   
75%        9859.000000              0.859000      4741.000000       81.000000   
max       34516.000000              3.397000     18484.000000      139.000000   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  \
count         10127.000000           10127.000000   
mean              0.712222               0.274894   
std               0.238086               0.275691   
min               0.000000               0.000000   
25%               0.582000               0.023000   
50%               0.702000               0.176000   
75%               0.818000               0.503000   
max               3.714000               0.999000   

       Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
count                                       10127.000000                                                                                    
mean                                            0.159997                                                                                    
std                                             0.365301                                                                                    
min                                             0.000008                                                                                    
25%                                             0.000099                                                                                    
50%                                             0.000181                                                                                    
75%                                             

# Data Cleaning

Based on our initial understanding of the data set we are working with, it is clear that some variables should be dropped - as they would not provide anything useful to our modeling or analysis. 

**CLIENTNUM** should be dropped, this is the client ID number - would not provide anything useful

**Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2** Notebook author noted that this column should be ignored and deleted 

**Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1** Notebook author noted that this column should be ignored and deleted

The remainder of the variables provide unique information on the customer that may influence the ultimate outcome of closing their credit card account. 

In [5]:
#Drop irrelevant columns 
credit_clean_df = credit_df.drop(columns=['CLIENTNUM',
        'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
        'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'])

In [6]:
#Check that the columns were dropped properly
credit_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Attrition_Flag            10127 non-null  object 
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           10127 non-null  object 
 5   Marital_Status            10127 non-null  object 
 6   Income_Category           10127 non-null  object 
 7   Card_Category             10127 non-null  object 
 8   Months_on_book            10127 non-null  int64  
 9   Total_Relationship_Count  10127 non-null  int64  
 10  Months_Inactive_12_mon    10127 non-null  int64  
 11  Contacts_Count_12_mon     10127 non-null  int64  
 12  Credit_Limit              10127 non-null  float64
 13  Total_Revolving_Bal       10127 non-null  int64  
 14  Avg_Op

There are now 20 columns with 10127 observations in total. No columns have any nulls present, so no cleaning is required on that end. There are multiple variables within this data set that are not numbers and need to be encoded as such so that they can be fed into our developed models and run properly. There are six different object variables.

Object variables that need to be cleaned / one-hot encoded / label encoded:
- Attrition Flag
- Gender
- Education Level
- Marital Status 
- Income Category 
- Card Category 

##### Attrition Flag Encoding
To view all values and their respective counts, use the .value_counts() function on the **Attrition_Flag** column. This way we will know the string values - giving us insight as to how we should encode them.

In [7]:
credit_clean_df['Attrition_Flag'].value_counts()

Existing Customer    8500
Attrited Customer    1627
Name: Attrition_Flag, dtype: int64

Redeclare the values in the column by mapping a dictionary of corresponding new values. In this case, encoding *existing customer* as 0 and *attrited customer* as 1 makes the most sense - we are looking to evaluate the factors that make a customer closer their account - model outputs will be most interpretable if the *attrited customers* are the positive case within this variable. 

In [8]:
credit_clean_df['Attrition_Flag'] = credit_clean_df['Attrition_Flag'].map({'Existing Customer': 0, 'Attrited Customer': 1})

##### Marital Status Encoding
Repeat the previous process to view all string values in the **Marital_Status** column. 

In [9]:
credit_clean_df['Marital_Status'].value_counts()

Married     4687
Single      3943
Unknown      749
Divorced     748
Name: Marital_Status, dtype: int64

Marital status is an object variable with no clear order that we can interpret numerically, so it makes the most sense to one-hot encode and create dummy variables for each value. 

##### Gender Encoding
Repeat the previous process to view all string values in the **Gender** column. 

In [10]:
credit_clean_df['Gender'].value_counts()

F    5358
M    4769
Name: Gender, dtype: int64

Gender also has no clear order that we can interpret numerically, so this should be encoded using one-hot encoding as well. Both **Marital_Status** and **Gender** can be encoded in the same line.

In [11]:
credit_clean_df = pd.get_dummies(credit_clean_df, columns = ['Marital_Status', 'Gender'], dtype=int)

##### Education Level Encoding
Repeat the previous process to view all string values in the **Education_Level** column. 

In [12]:
credit_clean_df['Education_Level'].value_counts()

Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: Education_Level, dtype: int64

Unlike Gender and Marital Status, Education level has a natural ranking and should be label encoded. The label encoder function, by default, ranks the object variable based on alphabetical value - this would be a potential cause for confusion when it comes time to interpret. To avoid this mislabeling we are manually encoding these labels.

Based on the values shown above, we are encoding as follows:
- 1 = Unknown
- 2 = Uneducated 
- 3 = High School
- 4 = College
- 5 = Graduate
- 6 = Post-Graduate
- 7 = Doctorate 

This way we can interpret higher values here to be higher education level reached. 

In [13]:
credit_clean_df['Education_Level'] = credit_clean_df['Education_Level'].apply(lambda x: ['Unknown', 'Uneducated', 'High School', 'College', 'Graduate', 'Post-Graduate', 'Doctorate'].index(x))
credit_clean_df['Education_Level'].value_counts()

4    3128
2    2013
0    1519
1    1487
3    1013
5     516
6     451
Name: Education_Level, dtype: int64

##### Income Category Encoding
Repeat the previous process to view all string values in the **Income_Category** column. 

In [14]:
credit_clean_df['Income_Category'].value_counts()

Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: Income_Category, dtype: int64

The same is true here as it was for **Education_Level**, this variable has a natural ranking. These values will need to be manually encoded as well.

Based on the values shown above, we are encoding as follows:
- 0 = Unknown
- 1 = Less than 40K 
- 2 = 40K - 60K
- 3 = 60K - 80K
- 4 = 80K - 120K
- 5 = 120K+

This way we can interpret higher values here to be higher income category of customer. 

In [15]:
credit_clean_df['Income_Category'] = credit_clean_df['Income_Category'].apply(lambda x: ['Unknown', 'Less than $40K', '$40K - $60K', '$60K - $80K', '$80K - $120K', '$120K +'].index(x))
credit_clean_df['Income_Category'].value_counts()

1    3561
2    1790
4    1535
3    1402
0    1112
5     727
Name: Income_Category, dtype: int64

##### Card Category Encoding
Repeat the previous process to view all string values in the **Card_Category** column. 

In [16]:
credit_clean_df['Card_Category'].value_counts()

Blue        9436
Silver       555
Gold         116
Platinum      20
Name: Card_Category, dtype: int64

This variable has a natural ranking just like the previous two. These values will need to be manually encoded as well.

Based on the values shown above, we are encoding as follows:
- 0 = Blue
- 1 = Silver
- 2 = Gold
- 3 = Platinum

This way we can interpret higher values here to be higher card category / tier of customer. 

In [17]:
credit_clean_df['Card_Category'] = credit_clean_df['Card_Category'].apply(lambda x: ['Blue', 'Silver', 'Gold', 'Platinum'].index(x))
credit_clean_df['Card_Category'].value_counts()

0    9436
1     555
2     116
3      20
Name: Card_Category, dtype: int64

Evaluate the resulting data frame.

In [18]:
credit_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Attrition_Flag            10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Dependent_count           10127 non-null  int64  
 3   Education_Level           10127 non-null  int64  
 4   Income_Category           10127 non-null  int64  
 5   Card_Category             10127 non-null  int64  
 6   Months_on_book            10127 non-null  int64  
 7   Total_Relationship_Count  10127 non-null  int64  
 8   Months_Inactive_12_mon    10127 non-null  int64  
 9   Contacts_Count_12_mon     10127 non-null  int64  
 10  Credit_Limit              10127 non-null  float64
 11  Total_Revolving_Bal       10127 non-null  int64  
 12  Avg_Open_To_Buy           10127 non-null  float64
 13  Total_Amt_Chng_Q4_Q1      10127 non-null  float64
 14  Total_

Now there are 23 columns due to the one-hot encoding of **Gender** and **Marital_Status**. Also after encoding, there are still no nulls, and now there are no longer and variables that are *object* DType. Since our data set is now all numbers, we can use **credit_clean_df** to run statistical analysis and begin our modeling phase.

##### Evaluating 'Unknown' Values in Categorical Columns

After looking at the values of each of the categorical variables, another potential cleaning issue came to surface. Even though the data frame has no null values, there is still missing information. In the variables, **Marital_Status**, **Education_Level** and **Income_Category**, there are 'Unknown' values. If there are observations with multiple 'Unknown' values for multiple categories then it may be a good idea to drop these observations. Otherwise, 'Unknown' will stay encoded as the numerical value of 0 for **Income_Category** and **Education_Level**, and one-hot encoded for **Marital_Status**. 

Looking into the 'Unknown' values to see if there are any observations with multiple pieces of unknown data. For simplicity, we will use the original dataframe, **credit_df** (not encoded), to check the number of observations that have multiple 'Unknown' values.

In [26]:
credit_df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

In [21]:
#Create a copy dataframe 
unknown_2 = credit_df.copy()

#Checking to see how many observations have two values missing / 'unknown'
unknown_2['matching1'] = unknown_2.apply(lambda x: x.Marital_Status == x.Education_Level, axis=1)
unknown_2['matching2'] = unknown_2.apply(lambda x: x.Marital_Status == x.Income_Category, axis=1)
unknown_2['matching3'] = unknown_2.apply(lambda x: x.Income_Category == x.Education_Level, axis=1)

#Subset dataframe that returns values only if one of the matches is true
unknown_2 = unknown_2[(unknown_2.matching1 == True) | (unknown_2.matching2 == True) | (unknown_2.matching3 == True)]

#Return number of observations in the dataframe 
print(len(unknown_2))

327


There are 327 observations with at least 2 columns of 'Unknown' values. That is the equivalent of about 3.22% of the total observations in the main dataset.

How many of these observations are from existing customers vs. attrited customers?

In [22]:
#Use subset data frame in order to get the value counts of the attrition flag
print(unknown_2['Attrition_Flag'].value_counts())

Existing Customer    270
Attrited Customer     57
Name: Attrition_Flag, dtype: int64


Of the 327 observations that have at least 2 columns of 'Unknown' values, 57 of the are attrited customers. There are 1627 attrited customer observations, meaning 3.5% of our attrited customer observations have 2+ 'Unknown' values.

How many observations have 3 'Unknown' values? 

In [23]:
unknown_3 = credit_df.copy()
unknown_3['matching'] = unknown_3.apply(lambda x: x.Marital_Status == x.Education_Level == x.Income_Category, axis=1)
unknown_3['matching'].value_counts()

False    10120
True         7
Name: matching, dtype: int64

There are 7 observations with 3 columns of 'Unknown' values. That is the equivalent of 0.07% of our data set, it is highly unlikely that such a small number would significantly skew any results.

In [24]:
#Use subset data frame in order to get the value counts of the attrition flag
unknown_3 = unknown_3.loc[unknown_3['matching'] == True]
print(unknown_3['Attrition_Flag'].value_counts())

Existing Customer    6
Attrited Customer    1
Name: Attrition_Flag, dtype: int64


There are only 7 cases where marital status, education level, and income category have the same 'Unknown' value. Of those 7, only 1 is an attrited customer - which is a positive sign that such a small number of our target observations have 3 unknown values. We will continue with the full dataset and not drop any observations. 

# Exploratory Data Analysis

Now that the data is cleaned and in a good state to work with, it is time to begin exploring relationships within the data. In this section we will leverage visualization packages and statistical analysis to gain a deeper understanding of the data. 

**credit_df** is the original data frame (no label encoding)

**credit_clean_df** is the data frame that will be used for modeling (label encoded)

In [25]:
#Correlation
credit_clean_df.corr()

Attrition_Flag  Customer_Age  Dependent_count  \
Attrition_Flag                  1.000000      0.018203         0.018991   
Customer_Age                    0.018203      1.000000        -0.122254   
Dependent_count                 0.018991     -0.122254         1.000000   
Education_Level                 0.008796     -0.002369         0.000472   
Income_Category                -0.013577      0.023508         0.066278   
Card_Category                   0.002354     -0.018235         0.030469   
Months_on_book                  0.013687      0.788912        -0.103062   
Total_Relationship_Count       -0.150005     -0.010931        -0.039076   
Months_Inactive_12_mon          0.152449      0.054361        -0.010768   
Contacts_Count_12_mon           0.204491     -0.018452        -0.040505   
Credit_Limit                   -0.023873      0.002476         0.068065   
Total_Revolving_Bal            -0.263053      0.014780        -0.002688   
Avg_Open_To_Buy                -0.000285      0.001151         0.068291   
Total_Amt_Chng_Q4_Q1           -0.131063     -0.062042        -0.035439   
Total_Trans_Amt                -0.168598     -0.046446         0.025046   
Total_Trans_Ct                 -0.371403     -0.067097         0.049912   
Total_Ct_Chng_Q4_Q1            -0.290054     -0.012143         0.011087   
Avg_Utilization_Ratio          -0.178410      0.007114        -0.037135   
Marital_Status_Divorced         0.000850     -0.042614         0.006697   
Marital_Status_Married         -0.023735      0.047364         0.014385   
Marital_Status_Single           0.019037     -0.011248        -0.040707   
Marital_Status_Unknown          0.008904     -0.026694         0.041743   
Gender_F                        0.037272      0.017312        -0.004563   
Gender_M                       -0.037272     -0.017312         0.004563   

                          Education_Level  Income_Category  Card_Category  \
Attrition_Flag                   0.008796        -0.013577       0.002354   
Customer_Age                    -0.002369         0.023508      -0.018235   
Dependent_count                  0.000472         0.066278       0.030469   
Education_Level                  1.000000        -0.011677       0.014989   
Income_Category                 -0.011677         1.000000       0.077326   
Card_Category                    0.014989         0.077326       1.000000   
Months_on_book                   0.006613         0.022122      -0.012535   
Total_Relationship_Count         0.000766        -0.003202      -0.094077   
Months_Inactive_12_mon           0.005761        -0.016310      -0.014629   
Contacts_Count_12_mon           -0.006280         0.023113      -0.000442   
Credit_Limit                    -0.002354         0.475972       0.492446   
Total_Revolving_Bal             -0.006800         0.034718       0.026304   
Avg_Open_To_Buy                 -0.001743         0.472760       0.489985   
Total_Amt_Chng_Q4_Q1            -0.010040         0.011352       0.007385   
Total_Trans_Amt                 -0.007460         0.019651       0.196003   
Total_Trans_Ct                  -0.004307        -0.054569       0.134275   
Total_Ct_Chng_Q4_Q1             -0.016692        -0.012657      -0.007261   
Avg_Utilization_Ratio           -0.001849        -0.246476      -0.198711   
Marital_Status_Divorced          0.005493        -0.008734      -0.004028   
Marital_Status_Married           0.014051         0.023808      -0.049254   
Marital_Status_Single           -0.015067        -0.024715       0.040783   
Marital_Status_Unknown          -0.004186         0.009417       0.021881   
Gender_F                         0.005087        -0.786608      -0.080093   
Gender_M                        -0.005087         0.786608       0.080093   

                          Months_on_book  Total_Relationship_Count  \
Attrition_Flag                  0.013687                 -0.150005   
Customer_Age                    0.788912                 -0.010931   
Dependent_coun

# Model Building

The purpose of the model building stage is to create a highly function classification model that can take in the numerous pieces of customer data and classify whether the customer is likely to churn / cancel their credit card. There are multiple routes that could generate a production grade model that could signficantly help the bank to understand what makes a customer likely to leave - in turn this could impact future marketing strategies, as well as allow them to be proactive and attempt to find methods of mitigating this issue among certain groups. 

## Logistic Regression

## Decision Tree

## Random Forest

## RF XGBoost

## RF AdaBoost

## Convolutional Neural Network (CNN)